src/main/kotlin/org/rust/lang/core/resolve/scope/RustResolveScope.kt

```diff
@@ -1,9 +1,256 @@
 package org.rust.lang.core.resolve.scope
 
-import org.rust.lang.core.psi.RustCompositeElement
-import org.rust.lang.core.psi.RustNamedElement
+import com.intellij.openapi.module.Module
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.psi.util.PsiUtilCore
+import org.rust.cargo.util.AutoInjectedCrates
+import org.rust.cargo.util.cargoProject
+import org.rust.cargo.util.getPsiFor
+import org.rust.cargo.util.preludeModule
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.impl.RustFile
+import org.rust.lang.core.psi.impl.mixin.isStarImport
+import org.rust.lang.core.psi.impl.rustMod
+import org.rust.lang.core.psi.util.module
 
 interface RustResolveScope : RustCompositeElement {
-    val declarations: Collection<RustNamedElement>
+    data class Context(
```

Comment:
> Yes, this is part of the resolve, and that's why it's in the org.rust.lang.core.resolve.scope package. RustResolveEngine does a bunch of other stuff, besides scopes, so by extracting scoping rules into a 'submodule' we can separate concerns. And this is possible because the dependency between scopes and RRE is unidirectional.

Primary goal of resolve engine -- is dealing with scoping rules. All the stuff related to `Entry`s is an _**implementation** detail_ and should not be exposed beyound the `RRE`.

Having the property of both completion & resolving engine using the same scoping infra is a nice thing, though it should be achived the other way -- completion engine should reference resolve-engine.